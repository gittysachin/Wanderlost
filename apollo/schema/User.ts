import { objectType, enumType, extendType, stringArg, idArg, arg } from '@nexus/schema'
import bcrypt from 'bcryptjs'
import jwt from 'jsonwebtoken'
import sgMail from '@sendgrid/mail'
import { promisify } from 'util'
import { randomBytes } from 'crypto'
import { setCookie } from '../cookies'

export const User = objectType({
  name: 'User',
  definition(t) {
    t.id('id')
    t.string('name')
    t.string('email')
    t.string('password')
    t.string('resetToken', { nullable: true })
    t.float('resetTokenExpiry', { nullable: true })
    t.field('permissions', {
      type: 'Permission',
      list: [false],
      nullable: true,
    })
    t.list.field('cart', { type: 'CartItem' })
    t.field('orders', {
      type: 'OrderItem',
      list: [false],
      nullable: true,
    })
  },
})

export const UserInput = objectType({
  name: 'UserInput',
  definition(t) {
    t.string('email')
    t.string('password')
    t.string('name')
  },
})

export const SuccessMessage = objectType({
  name: 'SuccessMessage',
  definition(t) {
    t.string('message')
  },
})

export const Permission = enumType({
  name: 'Permission',
  members: ['ADMIN', 'ITEMCREATE', 'ITEMDELETE', 'ITEMUPDATE', 'PERMISSIONUPDATE', 'USER'],
})

export const UserQuery = extendType({
  type: 'Query',
  definition(t) {
    t.field('user', {
      type: User,
      nullable: true,
      resolve(_root, _args, ctx) {
        if (!ctx.req.userId) {
          return null
        }
        return ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
          include: { cart: true },
        })
      },
    })

    t.field('users', {
      type: User,
      list: [false],
      nullable: true,
      resolve(_root, _args, ctx) {
        return ctx.prisma.user.findMany({
          include: { cart: true },
        })
      },
    })
  },
})

export const UserMutation = extendType({
  type: 'Mutation',
  definition(t) {
    t.field('signup', {
      type: User,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
        name: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        const user = await ctx.prisma.user.create({
          data: {
            email: args.email.toLocaleLowerCase(),
            name,
            password: await bcrypt.hash(args.password, 10),
            permissions: { set: ['USER', 'ITEMCREATE', 'ITEMDELETE', 'ITEMUPDATE'] },
          },
        })
        const token = jwt.sign({ userId: user.id }, process.env.APP_SECRET as string)
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365, // 1 year cookie
        })
        return user
      },
    })

    t.field('signin', {
      type: User,
      nullable: true,
      args: {
        email: stringArg({ required: true }),
        password: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if there is a user with that email
        const user = await ctx.prisma.user.findOne({ where: { email: args.email } })
        if (!user) {
          throw new Error(`No such user found for email ${args.email}`)
        }
        // 2. Check if their password is correct
        const valid = await bcrypt.compare(args.password, user.password)
        if (!valid) {
          throw new Error('Invalid Password!')
        }

        // 3. generate the JWT Token
        const token = jwt.sign({ userId: user.id }, process.env.APP_SECRET as string)

        // 4. Set the cookie with the token
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })
        // res.end(res.getHeader('Set-Cookie'))

        // 5. Return the user
        return user
      },
    })

    t.field('signout', {
      type: SuccessMessage,
      resolve(_root, _args, ctx) {
        setCookie(ctx.res, 'token', '', {})
        return { message: 'Goodbye!' }
      },
    })

    t.field('requestReset', {
      type: SuccessMessage,
      args: {
        email: stringArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        const user = await ctx.prisma.user.findOne({ where: { email: args.email } })
        if (!user) {
          throw new Error(`No user found for email: ${args.email}'`)
        }
        const randomBytesPromiseified = promisify(randomBytes)
        const resetToken = (await randomBytesPromiseified(20)).toString('hex')
        const resetTokenExpiry = Date.now() + 3600000 // 1 hour from now
        await ctx.prisma.user.update({
          where: { email: args.email },
          data: { resetToken, resetTokenExpiry },
        })
        const makeANiceEmail = (text: string) => `
          <div classname="email" style="
            border: 1px solid black;
            padding: 20px;
            font-family: sans-serif;
            line-height: 2;
            font-size: 20px;
          ">
            <h2>Hello there</>
            <p>${text}</p>
      
            <p>ðŸ˜˜, Davy Hausser</p>
          </div?
        `

        // TODO: make a nice email with functional link
        const msg = {
          to: user.email,
          from: 'davy.hausser@icloud.com',
          subject: 'Your Password Reset Token',
          text: 'and easy to do anywhere, even with Node.js',
          html: `<strong>${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}</strong>`,
          // html: makeANiceEmail(`Your Password Reset Token is here!
          // \n\n
          // <a href="${process.env.FRONTEND_URL}/reset?resetToken=${resetToken}">Click Here to Reset</a>`),
        }
        sgMail.setApiKey(process.env.SENDGRID_API_KEY)
        sgMail.send(msg)
        return { message: 'Thanks!' }
      },
    })

    t.field('resetPassword', {
      type: User,
      args: {
        resetToken: stringArg({ required: true }),
        password: stringArg({ required: true }),
        confirmPassword: stringArg({ required: true }),
      },
      async resolve(_root, args, ctx) {
        // 1. check if the passwords match
        if (args.password !== args.confirmPassword) {
          throw new Error("Yo Passwords don't match!")
        }
        // 2. check if its a legit reset token
        // 3. Check if its expired
        const [user] = await ctx.prisma.user.findMany({
          where: { resetToken: args.resetToken },
        })
        if (!user) {
          throw new Error('This token is either invalid or expired!')
        }
        // 4. Hash their new password
        const hash = await bcrypt.hash(args.password, 10)
        // 5. Save the new password to the user and remove old resetToken fields
        const updatedUser = await ctx.prisma.user.update({
          where: { email: user.email },
          data: {
            password: hash,
            resetToken: null,
            resetTokenExpiry: null,
          },
        })
        // 6. Generate JWT
        const token = jwt.sign({ userId: updatedUser.id }, process.env.APP_SECRET as string)
        // 7. Set the JWT cookie
        ctx.res.cookie('token', token, {
          httpOnly: true,
          maxAge: 1000 * 60 * 60 * 24 * 365,
        })
        // 8. return the new user
        return updatedUser
      },
    })

    t.field('updatePermissions', {
      type: User,
      nullable: true,
      args: {
        permissions: arg({
          type: 'Permission',
          list: [false],
        }),
        userId: idArg({ required: true }),
      },
      async resolve(_, args, ctx) {
        // 1. Check if they are logged in
        if (!ctx.req.userId) {
          throw new Error('You must be logged in!')
        }
        // 3. Check if they have permissions to do this
        const user = await ctx.prisma.user.findOne({
          where: { id: ctx.req.userId },
        })
        hasPermission(user, ['ADMIN', 'PERMISSIONUPDATE'])
        // 4. Update the permissions
        return ctx.prisma.user.update({
          data: {
            permissions: { set: args?.permissions },
          },
          where: {
            id: ctx.req.userId,
          },
        })
      },
    })
  },
})

const hasPermission = (user: any, permissionsNeeded: any) => {
  const matchedPermissions = user.permissions.filter((permissionTheyHave: any) =>
    permissionsNeeded.includes(permissionTheyHave)
  )
  if (!matchedPermissions.length) {
    throw new Error(`You do not have sufficient permissions
      : ${permissionsNeeded}
      You Have:
      ${user.permissions}
      `)
  }
}
