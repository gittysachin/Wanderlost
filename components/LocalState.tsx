import { createContext, useState, useContext } from 'react'

interface Props {
  children: JSX.Element
}

const LocalStateContext = createContext({
  cartOpen: false,
  toggleCart: function () {
    return
  },
  closeCart: function () {
    return
  },
  openCart: function () {
    return
  },
})
const LocalStateProvider = LocalStateContext.Provider

function CartStateProvider({ children }: Props) {
  const [cartOpen, setCartOpen] = useState(false)

  function toggleCart() {
    setCartOpen(!cartOpen)
  }

  function closeCart() {
    setCartOpen(false)
  }

  function openCart() {
    setCartOpen(true)
  }

  return (
    <LocalStateProvider
      value={{
        cartOpen,
        toggleCart,
        openCart,
        closeCart,
      }}
    >
      {children}
    </LocalStateProvider>
  )
}

function useCart() {
  return useContext(LocalStateContext)
}

export { CartStateProvider, LocalStateContext, useCart }
