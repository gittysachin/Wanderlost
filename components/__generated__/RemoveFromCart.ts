/* tslint:disable */
/* eslint-disable */
// @generated
// This file was automatically generated and should not be edited.

// ====================================================
// GraphQL mutation operation: RemoveFromCart
// ====================================================

export interface RemoveFromCart_deleteCartItem {
  __typename: 'CartItem'
  id: string
}

export interface RemoveFromCart {
  deleteCartItem: RemoveFromCart_deleteCartItem | null
}

export interface RemoveFromCartVariables {
  id: string
}
