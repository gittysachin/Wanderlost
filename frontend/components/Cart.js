import React from 'react'
import { useQuery, useMutation } from 'react-apollo-hooks'
import { gql } from 'apollo-boost'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import SickButton from './styles/SickButton'

export const LOCAL_STATE_QUERY = gql`
  query {
    cartOpen @client
  }
`

export const TOGGLE_CART_MUTATION = gql`
  mutation {
    toggleCart @client
  }
`

export default () => {
  const toggleCart = useMutation(TOGGLE_CART_MUTATION)
  const { data } = useQuery(LOCAL_STATE_QUERY)
  return (
    <CartStyles open={data.cartOpen}>
      <header>
        <CloseButton onClick={toggleCart} title="close">
          &times;
        </CloseButton>
        <Supreme>Your Cart</Supreme>
        <p>You have __ Items in your cart.</p>
      </header>

      <footer>
        <p>$10.10</p>
        <SickButton>Checkout</SickButton>
      </footer>
    </CartStyles>
  )
}
