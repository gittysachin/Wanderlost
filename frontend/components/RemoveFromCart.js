import { useMutation, gql } from '@apollo/client';
import styled from 'styled-components';
import PropTypes from 'prop-types';
import { CURRENT_USER_QUERY } from './User';

const REMOVE_FROM_CART_MUTATION = gql`
  mutation deleteCartItem($id: ID!) {
    deleteCartItem(id: $id) {
      id
    }
  }
`;

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${(props) => props.theme.red};
    cursor: pointer;
  }
`;

// This gets called as soon as we get a response back
// from the server after a mutation has been performed
function updateCart(cache, payload) {
  // 1. first read the cache
  const data = cache.readQuery({ query: CURRENT_USER_QUERY });
  // 2. remove that item from the cart
  const cartItemId = payload.data.deleteCartItem.id;
  const updatedCart = data.authenticatedUser.cart.filter(
    (cartItem) => cartItem.id !== cartItemId,
  );
  // 3. write it back to the cache
  cache.writeQuery({
    query: CURRENT_USER_QUERY,
    data: {
      ...data,
      authenticatedUser: {
        ...data.authenticatedUser,
        cart: updatedCart,
      },
    },
  });
}

function RemoveFromCart({ id }) {
  const [removeFromCart, { loading }] = useMutation(REMOVE_FROM_CART_MUTATION, {
    variables: { id },
    update: updateCart,
    optimisticResponse: {
      __typename: 'Mutation',
      deleteCartItem: {
        __typename: 'CartItem',
        id,
      },
    },
  });
  return (
    <BigButton
      disabled={loading}
      onClick={() => {
        // eslint-disable-next-line no-alert
        removeFromCart().catch((err) => alert(err.message));
      }}
      title="Delete Item"
    >
      &times;
    </BigButton>
  );
}

RemoveFromCart.propTypes = {
  id: PropTypes.string.isRequired,
};

export default RemoveFromCart;
export { REMOVE_FROM_CART_MUTATION };
