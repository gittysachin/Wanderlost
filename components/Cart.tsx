import { useCart } from './LocalState'
import { useUser } from './User'
import CartStyles from './styles/CartStyles'
import Supreme from './styles/Supreme'
import CloseButton from './styles/CloseButton'
import CartItem from './CartItem'
import calcTotalPrice from '../lib/calcTotalPrice'
import formatMoney from '../lib/formatMoney'
import Checkout from './Checkout'

function Cart() {
  let totalItems = 0
  const user = useUser()
  const { cartOpen, toggleCart } = useCart()
  if (!user) return null

  totalItems = user.cart.reduce((acc, current) => {
    return (acc = acc + current.quantity)
  }, 0)

  return (
    <CartStyles open={cartOpen}>
      <header>
        <CloseButton onClick={toggleCart} title="close">
          &times;
        </CloseButton>
        <Supreme>{user.name}'s Cart</Supreme>
        <p>
          You Have {totalItems} Item{totalItems === 1 ? '' : 's'} in your cart.
        </p>
      </header>
      <ul>
        {user.cart.map((cartItem) => (
          <CartItem key={cartItem.id} cartItem={cartItem} />
        ))}
      </ul>
      {user.cart.length && (
        <footer>
          <p>{formatMoney(calcTotalPrice(user.cart))}</p>
          <Checkout />
        </footer>
      )}
    </CartStyles>
  )
}

export default Cart
