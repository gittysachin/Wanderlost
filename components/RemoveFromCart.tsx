import { useMutation, gql, ApolloCache } from '@apollo/client'
import styled from 'styled-components'
import { CURRENT_USER_QUERY } from './User'
import { GetCurrentUser } from './__generated__/GetCurrentUser'
import {
  RemoveFromCart as RemoveFromCartTypes,
  RemoveFromCartVariables,
  RemoveFromCart_deleteCartItem,
} from './__generated__/RemoveFromCart'

const REMOVE_FROM_CART_MUTATION = gql`
  mutation RemoveFromCart($id: ID!) {
    deleteCartItem(id: $id) {
      id
    }
  }
`

const BigButton = styled.button`
  font-size: 3rem;
  background: none;
  border: 0;
  &:hover {
    color: ${(props) => props.theme.red};
    cursor: pointer;
  }
`

// This gets called as soon as we get a response back
// from the server after a mutation has been performed
function updateCart(
  cache: ApolloCache<RemoveFromCartTypes>,
  payload: { data?: RemoveFromCartTypes | null | undefined }
) {
  // 1. first read the cache
  const data = cache.readQuery<GetCurrentUser>({ query: CURRENT_USER_QUERY })
  if (!data?.user) return null
  // 2. remove that item from the cart
  const cartItemId = payload.data?.deleteCartItem?.id
  const updatedCart = data.user.cart.filter(
    (cartItem: RemoveFromCart_deleteCartItem) => cartItem.id !== cartItemId
  )

  // 3. write it back to the cache
  cache.writeQuery({
    query: CURRENT_USER_QUERY,
    data: {
      ...data,
      user: {
        ...data.user,
        cart: updatedCart,
      },
    },
  })
}

function RemoveFromCart({ id }: RemoveFromCartVariables) {
  const [removeFromCart, { loading }] = useMutation<RemoveFromCartTypes, RemoveFromCartVariables>(
    REMOVE_FROM_CART_MUTATION,
    {
      variables: { id },
      update: updateCart,
      optimisticResponse: {
        deleteCartItem: {
          __typename: 'CartItem',
          id,
        },
      },
    }
  )
  return (
    <BigButton
      disabled={loading}
      onClick={() => {
        removeFromCart().catch((err) => alert(err.message))
      }}
      title="Delete Item"
    >
      &times;
    </BigButton>
  )
}

export default RemoveFromCart
export { REMOVE_FROM_CART_MUTATION }
