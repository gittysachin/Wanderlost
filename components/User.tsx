import { useQuery, gql } from '@apollo/client'
import { GetCurrentUser } from './__generated__/GetCurrentUser'

const CURRENT_USER_QUERY = gql`
  query GetCurrentUser {
    user {
      id
      email
      name
      permissions
      cart {
        id
        quantity
        item {
          id
          price
          image
          title
          description
        }
      }
    }
  }
`

function useUser() {
  const { data } = useQuery<GetCurrentUser>(CURRENT_USER_QUERY)
  if (data) {
    return data.user
  }
  return null
}

export { CURRENT_USER_QUERY, useUser }
