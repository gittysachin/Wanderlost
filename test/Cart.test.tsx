import {
  render,
  waitFor,
  userEvent,
  fakeUser,
  cleanup,
  fakeCartItem,
} from '../lib/test-utils'

import { Cart } from '../components/Cart'
import { CURRENT_USER_QUERY } from '../components/User'
import { REMOVE_FROM_CART_MUTATION } from '../components/RemoveFromCart'
import { CartStateProvider } from '../components/LocalState'

const user = {
  ...fakeUser(),
  cart: [fakeCartItem('omg123'), fakeCartItem('abc123')],
}

const mocks = [
  {
    request: { query: CURRENT_USER_QUERY },
    result: {
      data: {
        user,
      },
    },
  },
  {
    request: { query: REMOVE_FROM_CART_MUTATION, variables: { id: 'omg123' } },
    result: {
      data: {
        deleteCartItem: { id: 'omg123' },
      },
    },
  },
]

describe('<Cart/>', () => {
  afterEach(cleanup)

  it('renders and matches the snapshot', async () => {
    const { container, getByText } = render(
      <CartStateProvider>
        <Cart />
      </CartStateProvider>,
      { mocks, addTypename: false }
    )

    await waitFor(() => getByText(`${user.name}'s Cart`))

    expect(getByText(`${user.name}'s Cart`)).toBeInTheDocument()
    expect(container).toMatchSnapshot()
  })
})

describe('<RemoveFromCart/>', () => {
  it('udpates when an item is removed', async () => {
    const { container, findByTestId, getAllByTitle } = render(
      <CartStateProvider>
        <Cart />
      </CartStateProvider>,
      { mocks, addTypename: false }
    )

    await findByTestId('cart')
    expect(container).toHaveTextContent(/You have 2 items/i)
    const deleteButtons = getAllByTitle(/Delete Item/i)
    expect(deleteButtons).toHaveLength(2)

    userEvent.click(deleteButtons[0])

    await findByTestId('cart')
    const deleteButtonsUpdated = getAllByTitle(/Delete Item/i)
    expect(container).toHaveTextContent(/You have 1 item/i)
    expect(deleteButtonsUpdated).toHaveLength(1)
  })
})
