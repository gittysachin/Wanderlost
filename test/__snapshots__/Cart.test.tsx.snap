// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Cart/> renders and matches the snapshot 1`] = `
<div>
  <div
    class="sc-bdnylx jMyLIc"
    data-testid="cart"
  >
    <header>
      <button
        class="sc-dlnjPT cEBaNc"
        title="close"
      >
        ×
      </button>
      <h3
        class="sc-gtssRu egxXhP"
      >
        Davy's Cart
      </h3>
      <p>
        You Have 
        3
         Item
        s
         in your cart.
      </p>
    </header>
    <ul>
      <li
        class="sc-eCApGN ehOohY"
      >
        <img
          alt="dogs are best"
          src="dog-small.jpg"
          width="100"
        />
        <div
          class="cart-item-details"
        >
          <h3>
            dogs are best
          </h3>
          <p>
            $150
             - 
            <em>
              3
               × 
              $50
               Each
            </em>
          </p>
        </div>
        <button
          class="sc-hKFyIo hjhEMf"
          title="Delete Item"
        >
          ×
        </button>
      </li>
    </ul>
    <footer>
      <p>
        $150
      </p>
      <form
        class="sc-pNWxx bBiJhz"
      >
        
        <div />
        <button
          class="sc-jSFkmK dJUPrO"
          type="submit"
        >
          Pay
        </button>
      </form>
    </footer>
  </div>
</div>
`;
